File: day19/day19.cpp
#include <fstream>
#include <vector>
#include <unordered_map>
#include <fmt/format.h>

using namespace std;

struct Symbol
{
	bool terminal;
	int value;
};

using Production = vector<Symbol>;

class Grammar
{
public:
	Grammar() : patched(false), pos(0) {}

	bool is_valid(const string& txt, bool patched)
	{
		Grammar::patched = patched;
		pos = 0;
		return match(txt, 0) && pos == txt.size();
	}

	bool match_all(const string& txt, const Production& p)
	{
		for (auto& sym: p)
		{
			if (sym.terminal)
			{
				if (txt[pos] != sym.value)
				{
					return false;
				}
				pos++;
			}
			else if (patched && sym.value == 8)
			{
				if (!match(txt, 42))
				{
					return false;
				}
			}
			else if (patched && sym.value == 11)
			{
				if (!match(txt, 42))
				{
					return false;
				}
				int cnt = 0;
				while (match(txt, 42))
				{
					cnt++;
				}
				if (!match(txt, 31))
				{
					return false;
				}
				while (match(txt, 31))
				{
					cnt--;
				}
				if (cnt < 0)
				{
					return false;
				}
			}
			else if (!match(txt, sym.value))
			{
				return false;
			}
		}
		return true;
	}

	bool match(const string& txt, int pid)
	{
		if (pos >= txt.size())
			return false;

		size_t tmp = pos;
		for (auto& rule: productions[pid])
		{
			if (match_all(txt, rule))
			{
				return true;
			}
			// backtrack
			pos = tmp;
		}
		return false;
	}

private:
	unordered_map<int,vector<Production>> productions;
	bool patched;
	size_t pos;

	friend istream& operator>>(istream&, Grammar& g);
};

istream& operator>>(istream& input, Grammar& g)
{
	g.productions.clear();
	string line;
	while (getline(input, line) && line.size())
	{
		size_t pos = line.find(':');
		if (pos == string::npos)
		{
			continue;
		}

		int pid = stoi(line.substr(0, pos));
		pos++;
		size_t nextp;
		do
		{
			while (isspace(line[pos]))
				pos++;
			nextp = line.find('|', pos);

			Production p;
			size_t symstart = pos;
			size_t symnext;
			do
			{
				symnext = line.find(' ', symstart);
				if (symnext > nextp)
					break;

				Symbol r;
				if (isdigit(line[symstart]))
				{
					r.terminal = false;
					r.value = stoi(line.substr(symstart, symnext-symstart));
				}
				else
				{
					r.terminal = true;
					r.value = line[symstart+1];
				}
				p.emplace_back(move(r));
				symstart = symnext + 1;
			} while (symnext != string::npos);
			g.productions[pid].emplace_back(move(p));
			pos = nextp + 1;
		} while (nextp != string::npos);
	}
	if (g.productions.size())
	{
		input.clear();
	}
	return input;
}


int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	Grammar g;
	input >> g;
	if (input.fail())
	{
		input.close();
		fmt::print(stderr, "Cannot parse the grammar rules\n");
		return 1;
	}

	size_t part1 = 0;
	size_t part2 = 0;
	string line;
	while (getline(input, line))
	{
		part1 += g.is_valid(line, false);
		part2 += g.is_valid(line, true);
	}
	input.close();

	fmt::print("Part1: {}\nPart2: {}\n", part1, part2);
	return 0;
}

File: day19/day19.py
#!/usr/bin/env python3

import sys

class Symbol(object):
    def __init__(self, terminal, value):
        self.terminal = terminal
        self.value = value

class Grammar(object):
    def __init__(self):
        self.productions = {}
        self.patched = False

    def parse(self, txt):
        self.productions.clear()
        rows = []
        state = 0
        for line in txt.strip().split("\n"):
            if state == 0:
                if line.strip() == "":
                    state = 1
                    continue

                pid, rest = line.split(": ")
                pid = int(pid)
                lst = []
                for r in rest.split(" | "):
                    if r[0] != "\"":
                        lst.append([Symbol(False, int(x)) for x in r.split(" ")])
                    else:
                        lst.append([Symbol(True, r[1])])
                    self.productions[pid] = lst
            else:
                rows.append(line)

        return rows

    def count_valid_rows(self, rows, patched):
        self.patched = patched
        return sum(1 for row in rows if self.is_valid(row))

    def is_valid(self, txt):
        self.pos = 0
        self.txt = txt
        return self.match(0) and self.pos == len(self.txt)

    def match_all(self, symbols):
        for s in symbols:
            if s.terminal:
                if self.txt[self.pos] != s.value:
                    return False
                else:
                    self.pos += 1
            elif self.patched and s.value == 8:
                if not self.match(42):
                    return False
            elif self.patched and s.value == 11:
                if not self.match(42):
                    return False
                cnt = 0
                while self.match(42):
                    cnt += 1
                if not self.match(31):
                    return False
                while self.match(31):
                    cnt -= 1
                if cnt < 0:
                    return False
            elif not self.match(s.value):
                return False

        return True

    def match(self, pid):
        if self.pos == len(self.txt):
            return False

        tmp = self.pos
        for rule in self.productions[pid]:
            if self.match_all(rule):
                return True
            # backtrack and start another rule
            self.pos = tmp

        return False

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    g = Grammar()
    rows = g.parse(txt)
    print("Part1:", g.count_valid_rows(rows, False))
    print("Part2:", g.count_valid_rows(rows, True))
